src/protocol.rs

```diff
@@ -1,42 +1,541 @@
 use std::net::{Ipv4Addr, Ipv6Addr};
 use std::str::FromStr;
 use std::convert::From;
+use std::fmt;
+use std::hash;
+use std::io;
 use std::io::Cursor;
+use std::io::Read;
+
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
-use cid::Cid;
-use integer_encoding::VarIntWriter;
+use cid;
+use integer_encoding::{VarIntReader, VarIntWriter};
 
 use {Result, Error};
 
+
 ///! # Protocol
 ///!
 ///! A type to describe the possible protocol used in a
 ///! Multiaddr.
 
-macro_rules! build_protocol_enum {
-    {$( $val:expr => $var:ident: $alph:expr, $size:expr, )*} => {
+
+/// Single multiaddress segment with its attached data
+pub trait AddressSegment : fmt::Display + ToString {
+    const STREAM_LENGTH: usize = 0;
+
+    fn protocol(&self) -> Protocol;
+
+    /// Read address segment data from stream
+    ///
+    /// If the address segment does expect any additional data, then
+    /// no data will be read and an empty struct will be created.
+    fn from_stream(stream: &mut io::Read) -> Result<Self> where Self: Sized;
+
+    /// Generate the canonical binary representation of the contents of this
+    /// address segment
+    ///
+    /// In order to obtain a human-readable string representation use
+    /// `ToString.to_string` instead.
+    fn to_stream(&self, stream: &mut io::Write) -> io::Result<()>;
+}
+
+impl<T> From<T> for Protocol where T: AddressSegment {
+    fn from(addr: T) -> Protocol {
+        addr.protocol()
+    }
+}
+
+
+
+/// A trait for reading any kind of address segment from a byte stream
+pub trait AddressSegmentReaderExt<T: AddressSegment> {
+    fn read_addr(&mut self) -> io::Result<T>;
+}
+impl<T: AddressSegment, R: io::Read> AddressSegmentReaderExt<T> for R {
+    fn read_addr(&mut self) -> io::Result<T> {
+        T::from_stream(self).map_err(|err: Error| {
+            io::Error::new(io::ErrorKind::InvalidData, err)
+        })
+    }
+}
+
+
+/// A trait for writing any kind of address segment to a byte stream
+pub trait AddressSegmentWriterExt<T: AddressSegment> {
+    fn write_addr(&mut self, addr: &T) -> io::Result<()>;
+}
+impl<T: AddressSegment, W: io::Write> AddressSegmentWriterExt<T> for W {
+    fn write_addr(&mut self, addr: &T) -> io::Result<()> {
+        addr.to_stream(self)
+    }
+}
+
+
+macro_rules! derive_for_wrapping_segment {
+    ( $name:ident ) => {};
+    ( $name:ident , $( $rest:tt )+ ) => {
+        derive_for_wrapping_segment!($name : $($rest)*);
+    };
+
+    ( $name:ident : Display $( $rest:tt )* ) => {
+        impl fmt::Display for $name {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                fmt::Display::fmt(&self.0, f)
+            }
+        }
+
+        derive_for_wrapping_segment!($name $($rest)*);
+    };
+
+    ( $name:ident : FromStr $( $rest:tt )* ) => {
+        impl FromStr for $name {
+            type Err = Error;
+
+            fn from_str(s: &str) -> Result<Self> {
+                Ok($name(FromStr::from_str(s)?))
```

Comment:
it's more idiomatic to write `parse()` instead of `FromStr::from_str`

Reply:
IMHO using `FromStr::from_str` here makes it more obvious that we are simply delegating the call to the inner type (otherwise I would prefer `.parse()` as well).

Reply:
makes sense! ;)